### DISCUSSION TITLE
Top asks from the Rules SIG of the Bazel team

### DISCUSSION THREAD

**Discussion by alexeagle** [bazel-contrib/SIG-rules-authors#24] [Ideas]
Let's keep a stack ranked list of things we wish the core team had time for.

- [ ] https://github.com/bazelbuild/bazel/issues/8830
- [ ] https://github.com/bazelbuild/bazel/issues/12691
- [ ] is rules_java dead? https://github.com/bazel-contrib/SIG-rules-authors/issues/27#issuecomment-986038016

**Comment by keith** [comment]
We've been keeping a list of these internally as well. I'm sure it would vary a bit given the types of things folks use bazel for. For example a lot of our P1s are Android issues

**Reply by brentleyjones** [reply]
On that list, https://github.com/bazelbuild/bazel/issues/12691 seems like it might be pretty applicable to most in this group.

**Reply by alexeagle** [reply]
https://github.com/bazelbuild/bazel/issues/8830 also - rulesets don't typically propagate tags like `no-remote-cache` so we need Bazel to do it

**Reply by keith** [reply]
Here's our internal list for reference and our internal priority for them

P2	https://github.com/bazelbuild/bazel/issues/10438
P2	https://github.com/bazelbuild/bazel/issues/13912
P2	https://github.com/bazelbuild/bazel/issues/13342
P2	https://github.com/bazelbuild/bazel/issues/13839
P2	https://github.com/bazelbuild/bazel/issues/13596
P2	https://github.com/bazelbuild/bazel/issues/13413
P2	https://github.com/bazelbuild/bazel/issues/12948
P2	https://github.com/bazelbuild/bazel/issues/12691
P2	https://github.com/bazelbuild/bazel/issues/12049
P2	https://github.com/bazelbuild/bazel/issues/10642
P3	https://github.com/bazelbuild/bazel/issues/8902
P3	https://github.com/bazelbuild/bazel/issues/14008
P3	https://github.com/bazelbuild/bazel/issues/14209

**Comment by alexeagle** [comment]
A broad topic: the bazel repo is home to the top-level "getting started" docs for each language, but the rulesets aren't maintained there.

JS: 2 years out of date, and no one should follow these docs:
https://github.com/bazelbuild/bazel/blob/master/site/docs/bazel-and-javascript.md

@keith are the docs linked from here still accurate? https://github.com/bazelbuild/bazel/blob/master/site/docs/bazel-and-apple.md

I think we may want to take this responsibility away from the Bazel docs site, given how hard it is to get review+merge of PRs to the bazelbuild/bazel repo. I pretty much gave up on helping to maintain those docs after waiting about five months on a trivial one-line change in https://github.com/bazelbuild/bazel/pull/13382

@aiuto @philwo do you have any thoughts about the future of language-specific docs on docs.bazel.build? Does your tech writing team plan to do a pass over these at any point?

**Reply by keith** [reply]
Looks like the apple ones are pretty outdated and incomplete. They wouldn't work with a modern version of bazel

**Reply by alexeagle** [reply]
I got a review and merge pretty quickly to remove the outdated JS page https://github.com/bazelbuild/bazel/pull/14366
Maybe just send one of those?

**Reply by keith** [reply]
Yea I'm pretty confident it would work, but I also wonder at this point if these examples would be better off living with the rulesets 😬

**Reply by alexeagle** [reply]
yes, that was my reasoning for removing it from bazel.build site.
https://bazelbuild.github.io/rules_nodejs/ ought to be strictly better in terms of content.
Do you already have a high-level landing page for Apple/Bazel ?

**Reply by keith** [reply]
Nope, just the readmes, there isn't really a high level tutorial at all today

**Reply by alexeagle** [reply]
Oh no... any ideas who has the knowledge needed to write it? I suppose we can use some of our SIG funding to pay someone for their time.

**Reply by keith** [reply]
It would likely have to be one of the maintainers of those repos now, or someone from google

**Comment by brentleyjones** [comment]
How about https://github.com/bazelbuild/bazel/issues/14327? This is blocking rules_swift from accepting a whole slew of upstream changes.

**Comment by alexeagle** [comment]
Doing some fresh brainstorming here, and at a bit more ambitious level than what the Bazel team has historically been staffed for.

1. Skyframe. Give a way to determine affected targets given some VCS delta. Lots of companies try to use this for test selection, but the current best answer https://github.com/Tinder/bazel-diff is not correct enough (errors can slip through and be misattributed to subsequent changes when the cache is busted). Make this thing a service which is fast enough even for larger workspaces. It's in other tools! https://monorepo.tools/#detecting-affected-projects-packages
2. Engage with community PRs. Dedicate real DevRel time to shepherding changes through. Currently most companies give up on contributing back and just fork stuff.
3. Tech writing. Danielle's team has been junior and churning on the move to developers.google.com. In the meantime Bazel core docs are in really bad shape, and the ruleset guides are years out-of-date.
4. Catalog community rulesets (or contribute to the SIG so we can do it) - Sven has been trying to track which repos under bazelbuild org are abandoned. There's nothing to help users know whether they can rely on something, even if Google appears to own it. Some rulesets are on life support like rules_docker
5. Other "devrel" stuff like Jin used to do. More activity on the Bazel blog highlighting the community for example. There was a push to groom the issue backlog on bazelbuild/bazel but not on all the other repos. Probably things I don't even know that DevRel does, like Angular DevRel spent lots of time on.
6. Treat google3 as one Bazel user. For example, "incompatible" flag flips are currently gated on them becoming the g3 default, rather than just adding to the Google global blazerc and allowing the rest of the community to start benefiting from the fix.
7. Own the bazel-central-registry content in more cases rather than waiting for the community. For example, make sure that all bazelbuild/* rulesets have their latest version mirrored to the registry.

**Reply by Wyverald** [reply]
This is a great list. My personal feeling (as an engineer on the Bazel team) is that I'd love all of this too, but we're just too understaffed.

> Like Bazel 5.0 came out without its blog post (at least so far) so the community wrote one (Brentley's summary). 

For the record, there _is_ a blog post (https://blog.bazel.build/2022/01/19/bazel-5.0.html) and it was out on the same day the binaries were published. So it's not like we're not trying :). With that said, [Brentley's blog post](https://www.buildbuddy.io/blog/whats-new-in-bazel-5-0) really _is_ awesome.

> Own the bazel-central-registry content in more cases rather than waiting for the community.

This is not really realistic with our team size. Also, I wonder why you think this is necessary -- if the rules can (should?) be owned by the community, why not the module definitions in the registry?

**Reply by alexeagle** [reply]
Oops on the blog post! I think it just wasn't published at the time Brentley posted his on slack.

Yeah this list is only feasible to consider if there were more resources.

**Reply by alexeagle** [reply]
The registry is a central mirror of what's happening out in the rulesets. And from end-user perspective, if their MODULE.bazel file doesn't work, it's not really germane whose fault that is - it's part of the Bazel user experience. I agree it would be a scope increase for the Bazel team, but since it's on the boundary between Bazel core and the rulesets, I think that's one example of where a scope increase would make sense.

**Reply by brentleyjones** [reply]
> Treat google3 as one Bazel user. For example, "incompatible" flag flips are currently gated on them becoming the g3 default, rather than just adding to the Google global blazerc and allowing the rest of the community to start benefiting from the fix.

This please. 1000 times this.

**Reply by keith** [reply]
> Treat google3 as one Bazel user. For example, "incompatible" flag flips are currently gated on them becoming the g3 default, rather than just adding to the Google global blazerc and allowing the rest of the community to start benefiting from the fix.

As discussed in our last meeting I'm actively working on the doc for this proposal

**Reply by meisterT** [reply]
Out of curiosity: can you give examples of flags which you really liked to have flipped earlier but have been delayed by waiting for Google?

**Reply by zachgrayio** [reply]
yeah definitely point 1 here imo. close to 100% of the companies we've worked with have this completely wrong when we come in, even if they're fairly mature otherwise. 

some docs and a fixed up version of this [old script](https://github.com/bazelbuild/bazel/blob/master/scripts/ci/ci.sh) in bazelbuild/bazel (is this even in use anymore?)  might be good enough as a starting point :laughing: though I guess thats something anyone here can do. but yeah, a `bazel affected` or `bazel ci` would be pretty awesome in the long term

**Reply by keith** [reply]
Here's the doc mentioned above https://docs.google.com/document/d/1-OTYIXWX-600MMAnXKTuF_5u-F6GRbkHobKLBrzPizU/edit# I'm hoping to discuss in the meeting today as well

**Comment by aiuto** [comment]
To give some more context about flags for people who were not in that
meeting.

We came up with a concept that Bazel should have a set of "recommend"
flags. These are the vision for what the next LTS release will have, and
what you would recommend to someone starting a fresh project today.,  This
is roughly the set of incompatible flags we intend to flip in the next
release + maybe some experimental flags.
